meta-facebook/yv35-gl/src/platform/plat_class.c

```diff
@@ -40,7 +46,232 @@ CARD_STATUS get_1ou_status()
+	msg.target_addr = SB_CPLD_ADDR;
+	msg.tx_len = 1;
+	msg.rx_len = 1;
+	msg.data[0] = SB_CPLD_REG_BOARD_CLASS_TYPE;
+	status = i2c_master_read(&msg, retry);
+	if (status)
+		LOG_ERR("failed to get class type from cpld, ret %d", status);
+	class_type = msg.data[0];
+	_1ou_status.present = ((class_type & BIT(2)) ? false : true);
+	_2ou_status.present = ((class_type & BIT(3)) ? false : true);
+
+	/* BIC judges the 1OU card type according the ADC-6(0-based) voltage.
+	 * The 1OU card type is
+	 *  - "SI test board" if voltage is lower than 0.3V
+	 *  - "Expansion with 6 M.2" if the voltage is 0.5V(+/- 5%)
+	 *  - "Rainbow falls ( CXL with 4 DDR4 DIMMs)" if the voltage is 0.75V(+/- 5%)
+	 *  - "Vernal falls (4 E1S, with TI chip)" if the voltage is 1.0V(+/- 5%)
+	 *  - "Waimano falls (CXL with 2 DDR4 DIMMs+2 E1S)" if the voltage is 1.26V(+/- 5%)
+	 *  - "Expansion with NIC" if the voltage is 1.5V(+/- 5%)
+	 * And then, BIC sets the type to CL CPLD register(slave address 21h, register 09h)
+	 * CPLD register 09h - 1OU Card Detection
+	 *  - 00h: 1OU SI test card
+	 *  - 01h: Expansion with 6 M.2
+	 *  - 02h: Rainbow falls ( CXL with 4 DDR4 DIMMs)
+	 *  - 03h: Vernal falls (with TI chip)
+	 *  - 04h: Vernal falls (with AST1030 chip)
+	 *  - 05h: Kahuna Falls
+	 *  - 06h: Waimano falls (CXL with 2 DDR4 DIMMs+2 E1S)
+	 *  - 07h: Expansion with NIC
+	 */
+	if (_1ou_status.present) {
+		float voltage;
+		bool success = get_adc_voltage(CHANNEL_6, &voltage);
+		if (success) {
+			for (int cnt = 0; cnt < ARRAY_SIZE(_1ou_card_mapping_table); cnt++) {
+				float typical_voltage = _1ou_card_mapping_table[cnt].voltage;
+				switch (_1ou_card_mapping_table[cnt].condition) {
+				case LOWER:
+					if (voltage <= typical_voltage) {
+						_1ou_status.card_type =
+							_1ou_card_mapping_table[cnt].card_type;
+					}
+					break;
+				case HIGHER:
+					if (voltage >= typical_voltage) {
+						_1ou_status.card_type =
+							_1ou_card_mapping_table[cnt].card_type;
+					}
+					break;
+				case RANGE:
+					if ((voltage >
+					     typical_voltage - (typical_voltage * 0.05)) &&
+					    (voltage <
+					     typical_voltage + (typical_voltage * 0.05))) {
+						_1ou_status.card_type =
+							_1ou_card_mapping_table[cnt].card_type;
+					}
+					break;
+				default:
+					LOG_ERR("unknown condition 0x%x",
+					       _1ou_card_mapping_table[cnt].condition);
+					break;
+				}
+
+				if (_1ou_status.card_type != TYPE_1OU_UNKNOWN) {
+					msg.tx_len = 2;
+					msg.rx_len = 0;
+					msg.data[0] = SB_CPLD_REG_1OU_CARD_DETECTION;
+					msg.data[1] = _1ou_status.card_type;
+					status = i2c_master_write(&msg, retry);
+					if (status)
+						LOG_ERR("failed to set 1ou card type, ret %d", status);
+					break;
+				}
+				if ((cnt == ARRAY_SIZE(_1ou_card_mapping_table)) &&
+				    (_1ou_status.card_type == TYPE_1OU_UNKNOWN)) {
+					LOG_ERR("unknown the 1OU card type, the voltage of ADC channel-6 is %fV",
+					       voltage);
+				}
+			}
+		}
+	}
+
+	/* TODO:
```

Comment:
Remove todo comment.

Reply:
Removed. 
Please review it again, thanks.